### DISCUSSION TITLE
Generaty output type like apollo

### DISCUSSION THREAD

**Discussion by JajaComp** [dotansimha/graphql-code-generator#7786] [Q&A] [answered]
Hi! Now i try use you tools for migration from apollo codegen. In my project i use >100 queries to graphql like this:
```
query Tasks {
  id
  name
  project {
    id
    name
  }
...
}
```
Apollo generate interface: 
```
export interface Tasks_tasks {
  id: Uuid;
  name: string;
  project: Tasks_tasks_project;
}
```
```
export interface Tasks_tasks_project {
  id: Uuid;
  name: string;
}
```
Can i generate some think like this with you tools? I use this interfaces for extends internal data models.

**Comment by n1ru4l** [comment, marked as answer]
It is not possible to generate interfaces instead of types.

If you want to extend them with an interface you can simply use the extends keyword.

```ts
type A = {
  a: number;
  b: number; 
}


interface B extends A {
  c: 1
}

function a(b: B) {
  b.a
  b.b
  b.c
}
```

[Example on TypeScript Playground](https://www.typescriptlang.org/play?ts=4.6.2&ssl=15&ssc=2&pln=1&pc=1#code/C4TwDgpgBAglC8UDeAoKUCGAuKA7ArgLYBGEATgNxpTE4EnkVQoC+K7AlrsOQGYYBjaACEoEAB49cAEwDOsZNQE4AjK3a98uAcA4B7XJgAUtKMICUi9MQB0GareIObA1kA)

**Reply by JDDoesDev** [reply]
> It is not possible to generate interfaces instead of types.

I have to disagree.  Apollo Codegen has been doing it for a while now and the code they use to generate what @JajaComp is mentioning can be found at https://github.com/apollographql/apollo-tooling/blob/be117f3140247b71b792a6b52b2ed26f2f76da02/packages/apollo-codegen-typescript/src/codeGeneration.ts#L253

It looks like things are being done there to create variants that make it easier to use the correct interfaces.  Is there a specific reason this wasn't included in graphql-code-generator?

**Reply by TheDutchCoder** [reply]
This is worrisome, because apollo is going to deprecate their codegen library and they force you to use an alternative like `graphql-code-generator`. Yet it doesn't provide the same functionality, so switching might be impossible for larger code bases.

Having no separate return types/interfaces on queries is a huge issue, because those types/interfaces need to be used to tell Vue (for example) what the type is of certain data or props. You can't just assign the result of the query and push it into a component, because it's just an Object and not typed.

**Reply by n1ru4l** [reply]
@TheDutchCoder You can use fragments for getting individual types that you can use for property typing.

**Reply by TheDutchCoder** [reply]
> @TheDutchCoder You can use fragments for getting individual types that you can use for property typing.

Unfortunately you can't because the fragment's type is not used in the returned value of the query.

**Reply by ardatan** [reply]
@TheDutchCoder If you use that fragment in that query, it will match the specific part of that operation so you can use the generated type of that fragment. If this still doesn't work you, please share an example on CodeSandbox or StackBlitz.

**Reply by TheDutchCoder** [reply]
> @TheDutchCoder If you use that fragment in that query, it will match the specific part of that operation so you can use the generated type of that fragment. If this still doesn't work you, please share an example on CodeSandbox or StackBlitz.

What do you mean exactly by "If you use that fragment in that query"?

The fragment itself is not very useful as the query could use the fragment + additional fields, those combined should generate a type/interface of its own, but graphql-code-generator doesn't do that.

I pass the query type to a function and the function uses the returns with the return type as part of the data (e.g. `response.data.getCollection` where `getCollection` should have a type/interface, but currently it's just an object.

I have a feeling I'm not explaining this well enough. I will add a code example in a minute.

**Reply by n1ru4l** [reply]
@TheDutchCoder Please provide an example on CodeSandbox. We cannot really understand the issue here.

**Reply by TheDutchCoder** [reply]
> @TheDutchCoder Please provide an example on CodeSandbox. We cannot really understand the issue here.

This is the query that's generated:
```
export type GetAccountQuery = { __typename?: 'Query', getAccount?: { __typename?: 'Account', id: string, username?: string | null, bio?: string | null, email?: string | null, banner?: any | null, icon?: any | null, type?: AccountTypes | null, verified?: boolean | null, social?: { __typename?: 'SocialLinks', twitter?: string | null, instagram?: string | null, website?: string | null } | null } | null };
```

And this is what I expect it to actually be:
```
export type GetAccountQuery = { __typename?: 'Query', getAccount?: GetAccountQueryAccount | null };

export type GetAccountQueryAccount = { __typename?: 'Account', id: string, username?: string | null, bio?: string | null, email?: string | null, banner?: any | null, icon?: any | null, type?: AccountTypes | null, verified?: boolean | null, social?: { __typename?: 'SocialLinks', twitter?: string | null, instagram?: string | null, website?: string | null } | null }
```

`GetAccountQueryAccount` would be the type of the response, so that if you assign to something (like a ref in Vue) you can make sure it's of the correct type.
